RFC20: Add Resource Set Specification RFC

dongahn opened the pull request:
This PR adds a new RFC that specifies the format 1 of resource set specification. This is discussed and agreed at https://github.com/flux-framework/rfc/issues/109#issuecomment-460469688.

dongahn commented:
As discussed at https://github.com/flux-framework/rfc/issues/109#issuecomment-460469688 as well as our last Thu meeting, I created a RFC20 draft for `R`. 

I have only added very basic keys for both `R_lite` as well as `scheduling`. This will further have to be optimized later. But I didn't want to add any premature optimization at this point.

@grondo, @SteVwonder, @garlick and @trws, let me know what you think.

dongahn commented:
CI failed because I didn't created a link to RFC20 in the top-level README.adoc.

dongahn commented:
OK. Forced a push. Hopefully this will make Travis happy.

SteVwonder commented on spec_20.adoc:
```diff
@@ -0,0 +1,156 @@
+keys: `version`, `execution` and `scheduling`. `R` SHALL be valid if and only
+if it contains the `version` key and either or both the `execution`
+and `scheduling` keys. The value of the `execution` key SHALL contain
+sufficient data for the execution subsystem CAN perform its
```
`CAN` isn't one of the formally defined terms in RFC 2119.  So I don't think it should be in all caps here.  I think `to perform` would also read a bit better.

dongahn replied:
I didn't realize this. Thanks.

SteVwonder commented on spec_20.adoc:
```diff
@@ -0,0 +1,156 @@
+* Users MAY manually write an `R` document for testing and debugging.
+* User-facing utilities that query a resource status (e.g., what
+  resources are available or idle, or what resources are allocated to a job)
+  CAN use an `R` document to extract this information;
```
`CAN` -> `MAY`?

grondo replied:
:+1:

dongahn replied:
Agreed.

SteVwonder commented on spec_20.adoc:
```diff
@@ -0,0 +1,156 @@
+
+==== Execution
+
+The value of the `execution` key SHALL contain exactly one key: `R_lite`. `R_lite`
```
Borrowing from RFC 14, it might be useful to add some language here along the lines of: `with other keys reserved for future extensions` (if that, indeed, is the intent).

SteVwonder commented on spell.en.pws:
```diff
@@ -384,3 +384,11 @@ FLUIDs
+deserialized
+deserializes
+JGF
+medadata
```
This shouldn't be added to the spell list.  `medadata` should actually be `metadata`

dongahn replied:
Opps.

grondo commented on spec_20.adoc:
```diff
@@ -0,0 +1,156 @@
+    *core*::: The `core` key contains a logical compute core IDs string.
+     It SHALL be a comma-separated list of core IDs (e.g., 1,3,4,5) or
+     hyphenated ranges for consecutive integer IDs (e.g., 1,3-5).
+
```
Do we want to add OPTIONAL `gpu` key here as well since current `R_lite` supports it?

dongahn replied:
Good thought. Agreed.

grondo commented on spec_20.adoc:
```diff
@@ -0,0 +1,156 @@
+sufficient data for the execution subsystem CAN perform its
+core tasks. The value of `scheduling` SHALL contain sufficient data
+for schedulers.
+
```
The `execution` or similar key should be required not optional (this will be the part of `R` used by flux-core). The `scheduling` key should be OPTIONAL as a key for scheduler-specific data.

In fact, maybe we should rename `execution` to `core`? (Just a thought)

dongahn replied:
Thanks.

Re: optional.

One counter example would be when the execution system releases a resource subset to the scheduler, this key shouldn't be necessary in such cases?

The reason I used `execution` was for a potential case where this section is extended like `R_lite++`.

But this should really be adjusted per your development trajectory though. I was thinking the execution system will evolve before being able to fully adapt to the API approach we agreed to do for the future. 

I may be wrong though.

grondo replied:
Good thoughts. I hadn't thought about how the execution service could create an R subset when it also contains a `scheduling` dictionary. Since that is supposed to be opaque I think it either has to leave that off and require the scheduler to be able to determine which exact resources are being released from R_lite only, or leave the entire `scheduling` section alone and pass it through.

Long term I think you and @trws are correct that it is better to use the one-section approach, but that change can come after this stop-gap spec.

For now can we kick the can down the road and assume the only resource subset than can be released is "by rank"?

Pragmatically, the reason `scheduling` should be optional is that when testing in `flux-core` the dummy scheduler will not use this section, and we still want to create an `R` that adheres to the spec.

dongahn replied:
Thanks.

> For now can we kick the can down the road and assume the only resource subset than can be released is "by rank"?

Yeah this seems right. Why don't we target this for now and see if we need to adjust our spec as we continue to develop the functionality...

> Pragmatically, the reason scheduling should be optional is that when testing in flux-core the dummy scheduler will not use this section, and we still want to create an R that adheres to the spec.

Agreed. 

BTW, my reason for making `execution` optional was the same. I wanted to create an `R` document for scheduler tests like the new reader with this key adsence.

In general, does the current spec meet your requirement or do you need to make the execution key still mandatory?

grondo replied:
Ah I see, I think it is fine if the `execution` key is optional for now, with the caveat that without `execution` key the `R` can't create a valid Flux job.

grondo commented on spec_20.adoc:
```diff
@@ -0,0 +1,156 @@
+The `graph` key SHALL conform to the latest version of the JSON Graph Format (JGF).
+Thus, its value is a dictionary with two keys, `nodes` and `edges`,
+that encode the resource vertices and edges as described in RFC 4.
+
```
In this version of `R`, the purpose of the `scheduling` dict was to an area for advanced schedulers to add data opaque to flux-core. So, the dict should really allow any keys, with `graph` reserved for your JGF spec here. (However, it seems unlikely there will be any other schedulers using this spec for now, so I'm fine with the current language)

dongahn replied:
True. But thinking was to make this spec beyound flux-core centric. 

In general there will be thr base scheduler data per RFC4 that needs to be shoved into the graph, I though it would be good specify them.

grondo replied:
Oh, yeah, sorry I think I wasn't clear. I only meant to change something like:

> The scheduling key contains a dictionary with a single key: graph

to allow other keys besides `graph`

I'm fine with specifying the `graph` key in this RFC.

dongahn replied:
Ah yes. I think that more defined keys should be added there at that level later. In addition other arbitrary should be allowed.

For eample I will probably need the `properties` key to have a properties map for edges. 

I just wanted to keep things simple and grow the spec as we develop this within our resource matching service.

For now, yes I think it makes sense to change 

> The scheduling key contains a dictionary with a single key: graph

to allow other arbitrary keys at this level.

garlick commented on spec_20.adoc:
```diff
@@ -0,0 +1,156 @@
+the execution subsystem then MUST be able to extract a per-rank,
+local resource set for every flux-broker rank in order to perform
+its work such as the containment, binding, mapping, and execution
+of program processes.
```
Suggestion:  IMHO the opening paragraph/section should provide human digestible background, and the precise RFC 2119 language should be reserved for requirements/goals and the format definition.  I find these RFCs easier to digest when there is some expository writing at the top, and some really precise "programmer instructions" further in, and depending on which hat one has on, one or the other can be somewhat glossed over.

For the intro, maybe something like:
> Flexible resource representation is an important Flux value proposition.  Resource _requests_ are part of Flux jobspec, described in RFC 16.  This RFC describes the format of concrete resource sets, referred to as _R_, constructed by the scheduler in response to a resource request.
>
> _R_ is input to the core execution system, which uses information expressed in _R_ to establish containment, binding, mapping, and execution of program tasks, apportioned across broker ranks.  As a parallel program terminates, the execution system releases shards of the original _R_, eventually adding up to its union, back to the scheduler.
>
> Finally, when Flux launches Flux, _R_ is passed down from enclosing to child Flux instance, where it primes the child scheduler with a block of allocatable resources.

dongahn replied:
This sounds good to. I will incorporate this.

garlick replied:
Oops:  `s/RFC 16/RFC 14/`

grondo commented on spec_20.adoc:
```diff
@@ -0,0 +1,156 @@
+
+==== Version
+
+The value of the `version` key SHALL contain an non-negative integer that
```
:+1:  Good catch!

dongahn replied:
Yes. Thanks.

dongahn commented on spec_20.adoc:
```diff
@@ -0,0 +1,156 @@
+* The program execution service emits a valid `R` document to release
+  a resource subset of an `R` to the instance scheduler.
+
+=== Resource Set Format Definition
```
Good catch. Unintended. Will add a section.

dongahn commented:
@SteVwonder, @grondo, @garlick: Okay, I believe I incorporate all of your comments. Please let me know if you want me to address more issues.

garlick commented on spec_20.adoc:
```diff
@@ -26,62 +26,69 @@ in this document are to be interpreted as described in RFC 2119.
+adding up to its union, back to the scheduler.
+Finally, when a Flux instance launches a child instance,
+_R_ is passed down from the enclosing to the child, where it primes
+the child scheduler with a block of allocatable resources.
```
Probably my bad but "passed down from the enclosing to the child" should probably read "passed down from the enclosing instance to the child instance".

garlick commented on spec_20.adoc:
```diff
@@ -26,62 +26,69 @@ in this document are to be interpreted as described in RFC 2119.
+_R_ is passed down from the enclosing to the child, where it primes
+the child scheduler with a block of allocatable resources.
+
+== Design Goal
```
Goal should be plural.

garlick commented on spec_20.adoc:
```diff
@@ -26,62 +26,69 @@ in this document are to be interpreted as described in RFC 2119.
-
-An `R` JSON document SHALL consist of a dictionary with three
-keys: `version`, `execution` and `scheduling`. `R` SHALL be valid if and only
+  MAY use an _R_ document to extract this information;
```
Is _R_ a "document"?  Maybe "object" is a better word choice.

dongahn replied:
Thanks. Since what an _R_ document is is defined later, I will use "object" here. Glad you mentioned this, though, as I struggled a bit with what _R_ actually defines from the previous revision. In this revision, I used the following definition. Hopefully it makes sense. 

- R is defined to be "the resource-set representation" 
- The R format: the format of this representation
- An R JSON document or in short an R document: a document that conforms to the format in JSON.

dongahn commented:
Addressed @grondo and @garlick's comments and squashed the commits. This should be ready to go in. Thanks everyone for reviewing it. It should give me a firm target to work on. Don't need to worry about other emit types...

grondo commented:
Thank you, @dongahn!

dongahn commented:
Thanks.
